src/functionlenses.jl

```diff
@@ -33,6 +38,37 @@ set(obj::Dict, lens::Union{typeof(keytype),typeof(valtype)}, T::Type) =
 set(obj::Type{<:Dict{<:Any,V}}, lens::typeof(keytype), ::Type{K}) where {K,V} = Dict{K,V}
 set(obj::Type{<:Dict{K}}, lens::typeof(valtype), ::Type{V}) where {K,V} = Dict{K,V}
 
+################################################################################
+##### array shapes
+################################################################################
+set(obj, ::typeof(size), v::Tuple) = reshape(obj, v)
+
+# set vec(): keep array shape and type, change its values
+function set(x::AbstractArray, ::typeof(vec), v::AbstractVector)
+    res = similar(x, eltype(v))
+    vec(res) .= v
+    res
+end
+
+# set reverse(): keep vector type, change its values
+function set(x::AbstractVector, ::typeof(reverse), v::AbstractVector)
+    res = similar(x, eltype(v))
+    res .= v
+    reverse!(res)
+    res
+end
+
+################################################################################
+##### ranges
+################################################################################
+set(r::AbstractRange, ::typeof(step), s) = range(first(r), last(r), step=s)
+set(r::AbstractRange, ::typeof(length), l) = range(first(r), last(r), length=l)
+set(r::AbstractRange, ::typeof(first), x) = range(x,  last(r), step=step(r))
+set(r::AbstractRange, ::typeof(last),  x) = range(first(r), x, step=step(r))
+set(r::AbstractUnitRange, ::typeof(first), x) = x:last(r)
+set(r::AbstractUnitRange, ::typeof(last),  x) = first(r):x
+set(r::Base.OneTo, ::typeof(last),  x) = Base.OneTo(x)
```

Comment:
Behavior is not obvious here for me. For instance one can change the first element of a range in two ways:
* Keep `length` and `step` but change `last`
* Keep `length` and `last` but change `step`

So I am not sure we should have these methods.

Reply:
Yeah these manipulations are useful. If we find a way to disambiguate I would be happy to have them.